File: server/src/models/donationCenter.js
const mongoose = require("mongoose");
const mongoosePaginate = require("mongoose-paginate");

const donationCenterSchema = new mongoose.Schema(
  {
    name: {
      type: String,
    },
    address: {
      type: {
        state: {
          type: String,
        },
        city: {
          type: String,
        },
        woreda: {
          type: String,
        },
      },
    },
    status: {
      type: String,
      enum: ["open", "closed"],
      default: "open",
    },
    location: {
      //GEO JSON Point
      type: {
        type: String,
        enum: ["Point"],
      },
      coordinates: {
        type: [Number],
        index: "2dsphere",
      },
      formattedAddress: String,
      street: String,
      city: String,
      state: String,
      zipcode: String,
      country: String,
    },
    timeSlot: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "TimeSlot",
    },
    createdBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    updatedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    totalDonations: {
      type: Number,
      default: 0,
    },
    isDeleted: {
      type: Boolean,
      default: false,
    },
  },
  {
    timestamps: true,
  }
);

donationCenterSchema.plugin(mongoosePaginate);

const DonationCenter = mongoose.model("DonationCenter", donationCenterSchema);

module.exports = DonationCenter;

File: client/lib/src/models/donation_center.dart
class DonationCenter {
  DonationCenter(
      {this.id,
      required this.name,
      this.createdBy,
      this.updatedBy,
      this.status,
      this.timeSlot,
      required this.address,
      this.totalDonations});
  String? id;
  String name;
  String? status;
  String? createdBy;
  String? updatedBy;
  Map? timeSlot;
  Map address;
  int? totalDonations;

  factory DonationCenter.fromJson(Map<String, dynamic> json) {
    print(json["totalDonations"].runtimeType);

    print(
        "*********************** DonationCenter fromJson ****************************");
    String? createdBy;
    String? updatedBy;
    // print(json["createdBy"]["_id"].toString());
    try {
      // account for different types of response
      createdBy = json["createdBy"]["_id"];
    } catch (e) {
      createdBy = json["createdBy"];
    }

    try {
      updatedBy = json["updatedBy"]["_id"];
    } catch (e) {
      updatedBy = json["updatedBy"];
    }

    return DonationCenter(
      id: json["_id"],
      name: json["name"],
      address: json["address"],
      timeSlot: json["timeSlot"],
      status: json["status"],
      createdBy: createdBy,
      // updatedBy: updatedBy,
      // totalDonations: json["totalDonations"]
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "_id": id,
      "name": name,
      "address": address,
      "timeSlot": timeSlot,
      "status": status,
      "createdBy": createdBy,
      "updatedBy": updatedBy,
      "totalDonations": totalDonations
    };
  }

  @override
  String toString() {
    // TODO: implement toString
    return """{\n
        _id: $id, \n
        name: $name,\n
        status: $status,\n
        address: $address,\n
        createdBy: $createdBy,\n
        UpdatedBy: $updatedBy,\n
        timeSlot: $timeSlot,\n
        totalDonations: $totalDonations,\n
    }""";
  }
}
